source/common/formatter/substitution_formatter.h

```diff
@@ -253,8 +255,192 @@ template <class FormatterContext> class FormatterBaseImpl : public FormatterBase
+      }
+    }
+  }
+
+  std::string formatWithContext(const FormatterContext& context,
+                                const StreamInfo::StreamInfo& info) const {
+    std::string log_line;
+    log_line.reserve(512);
+
+    std::string sanitize_buffer;
+
+    for (const auto& element : elements_) {
+      // 1. Handle the raw string element.
+      if (absl::holds_alternative<std::string>(element)) {
+        log_line.append(absl::get<std::string>(element));
+        continue;
+      }
+      ASSERT(absl::holds_alternative<Formatters>(element));
+
+      const auto& formatters = absl::get<Formatters>(element);
+      ASSERT(!formatters.empty());
+
+      // 2. Handle the formatter element with multiple providers or case
+      //    that value type needs not to be kept.
+      if (formatters.size() > 1 || !keep_value_type_) {
+        // Multiple providers forces string output.
+        log_line.push_back('"');
+        for (const auto& provider : formatters) {
+          const auto bit = provider->formatWithContext(context, info);
+          if (!bit.has_value()) {
+            log_line.append(empty_value_);
+            continue;
+          }
+          log_line.append(Json::sanitize(sanitize_buffer, bit.value()));
+        }
+        log_line.push_back('"');
+
+        continue;
+      }
+
+      // 4. Handle the formatter element with a single provider and value
+      //    type needs to be kept.
+      auto value = formatters[0]->formatValueWithContext(context, info);
+      switch (value.kind_case()) {
+      case ProtobufWkt::Value::KIND_NOT_SET:
+      case ProtobufWkt::Value::kNullValue:
+        log_line.append(JsonFormatBuilder::NullValue);
+        break;
+      case ProtobufWkt::Value::kNumberValue:
+        log_line.append(fmt::to_string(value.number_value()));
+        break;
+      case ProtobufWkt::Value::kStringValue:
+        log_line.push_back('"');
+        log_line.append(Json::sanitize(sanitize_buffer, value.string_value()));
+        log_line.push_back('"');
+        break;
+      case ProtobufWkt::Value::kBoolValue:
+        log_line.append(value.bool_value() ? JsonFormatBuilder::BoolTrue
+                                           : JsonFormatBuilder::BoolFalse);
+        break;
+      case ProtobufWkt::Value::kStructValue:
+      case ProtobufWkt::Value::kListValue:
+        // Convert the struct or list to string. This may result in a performance
+        // degradation. But We rarely hit this case.
+        // Basically only metadata or filter state formatter may hit this case.
+#ifdef ENVOY_ENABLE_YAML
+        absl::StatusOr<std::string> json_or =
+            MessageUtil::getJsonStringFromMessage(value, false, true);
+        if (json_or.ok()) {
+          log_line.append(json_or.value());
+        } else {
+          log_line.push_back('"');
+          log_line.append(Json::sanitize(sanitize_buffer, json_or.status().message()));
```

Comment:
this is all code we really should be using JsonStreamer for.

Reply:
Comment/explanation from another comment:

The core of whole design is to do as much as possible work when loading configuration. The whole JSON constructing is split to two parts.

The first part, when loading configuration, the keys, raw values, delimiters will be serialized as JSON string pieces (raw JSON strings) directly.
The second part, when flushing log entry, the output of substitution formatter (values that generated by substitution commands) will be sanitized at runtime.
An array (`FormatterElements`) is used to ensure these JSON string pieces and output of formatter will in the correct position. We only need to join them in order to construct the final complete JSON output.

This is why this implementation could get the comparable performance with TEXT formatter. The logic at runtime (core data paths) is almost same with TEXT formatter: joining raw string pieces and output of formatter providers.
This is also why the `JsonStreamer` couldn't be used here. The JsonStreamer works perfect to serialize a tree to JSON. But cannot help to convert an array like this to final JSON output:

```
- '{"text":"text","tmpl":'                                       # Element 0, raw JSON piece.
- '%START_TIME% %FLAG%'                                          # Element 1, formatter string, must be independent element.
- ',"number":2,"bool":true,"list":["list_raw_value",false,'      # Element 2, raw JSON piece.
- '%EMIT_TIME%'                                                  # Element 3, formatter string, must be independent element.
- '],"nested":{"text":"nested_text"}}'                           # Element 4, raw JSON piece.
```